File: code_cpp/BST/3_deleting_a_node_in_BST.cpp
#include<iostream>
#include<queue>
#include<bits/stdc++.h>
using namespace std;
// Constructing Binary Tree
// create node of tree 
struct Node{

    int key;
    Node *left;
    Node *right;
    Node(int k){
        key=k;
        left=NULL;
        right=NULL;
    }
};
// successor of a node
// used to find when we delete a node which have both side subtree
Node *getSuccessor(Node *curr)
{
    curr=curr->right;
    while(curr!=NULL && curr->left!=NULL)
    {
        curr=curr->left;
    }
    return curr;
}
Node *delNode(Node *root, int x)
{
    if(root==NULL) return root;
    if (root->key>x)
    {
        root->left=delNode(root->left,x);
    }
    else if (root->key<x)
    {
        root->right=delNode(root->left,x);
    }
    else    // Node found, which we will be deleting.
    {
        // to delete bottom node or node having left part null 
        // but have right part some subtree
        if(root->left==NULL)
        {
            Node *temp=root->right;
            delete(root);
            return temp;
        }
        // to delete bottom node or node having right part null 
        // but have left part some subtree
        else if (root->right==NULL)
        {
            Node *temp=root->left;
            delete(root);
            return temp;
        }
        else    // if deleting node which have left and right both part available
        {
            Node *succ=getSuccessor(root);
            root->key=succ->key;// replace deleting node with successor node
            root->right=delNode(root->right,succ->key); // delete successor node now which a leaf node
        }
        
        
    }
    return root;
    
    
    
}
// Search in Binary search tree: recursive approch---> time=O(log N)
bool search(Node *root, int x)
{
    // base case
    if(root==NULL)
        return false;
    else if(root->key==x)
        return true;
    else if(root->key>x)
        return search(root->left,x);
    else
        return search(root->right,x);
}
// traversal
void inorder(Node *root){
    if(root!=NULL){
        inorder(root->left);
        cout<<root->key<<" ";
        inorder(root->right);    
    }
} 

int main(){
    // Node *root=NULL;    // empty tree
    Node *root=new Node(15);
    root->left=new Node(5);
    root->right=new Node(20);
    root->left->left=new Node(3);
    root->right->right=new Node(80);
    root->right->left=new Node(18);
    root->right->left->left=new Node(16);
    // cout<<search(root,2)<<endl;
    int x=15;
	
	root=delNode(root,x);
	inorder(root);
   
    return 0;
}

File: code_java/self pased DSA/BST/3_deleting_a_node_in_BST.java
/**
 * Problem: Delete a node from a Binary Search Tree (BST)
 *
 * Description:
 * Given a BST and a key 'x', delete the node with the key value 'x' while maintaining
 * the BST properties. A BST is a binary tree where:
 * - The left subtree of a node contains only nodes with keys less than the node's key.
 * - The right subtree contains only nodes with keys greater than the node's key.
 *
 * Approach:
 * We use a recursive approach to delete a node from the BST.
 * - If the key is smaller than the current node, go to the left subtree.
 * - If the key is greater, go to the right subtree.
 * - If the key matches, handle three cases:
 *   1. Node has no left child: return right child.
 *   2. Node has no right child: return left child.
 *   3. Node has two children:
 *      - Find the inorder successor (smallest node in the right subtree).
 *      - Replace the current node’s key with the successor's key.
 *      - Recursively delete the successor node.
 *
 * Why this approach:
 * This recursive method preserves the BST properties while properly managing memory.
 * It efficiently handles all three deletion cases and uses the inorder successor
 * strategy to maintain correct order when the node has two children.
 *
 * This works because:
 * - Recursive calls ensure the correct subtree is modified.
 * - Replacing with inorder successor maintains the BST ordering.
 * - Leaf or one-child cases are resolved by directly reconnecting child nodes.
 */

import java.util.*;

class Node {
    int key;
    Node left, right;

    Node(int key) {
        this.key = key;
        this.left = null;
        this.right = null;
    }
}

public class BSTDelete {

    // Inorder successor: leftmost node in right subtree
    public static Node getSuccessor(Node curr) {
        curr = curr.right;
        while (curr != null && curr.left != null) {
            curr = curr.left;
        }
        return curr;
    }

    // Delete a node in BST
    public static Node delNode(Node root, int x) {
        if (root == null)
            return root;

        if (x < root.key)
            root.left = delNode(root.left, x);
        else if (x > root.key)
            root.right = delNode(root.right, x);
        else {
            // Node with only one child or no child
            if (root.left == null) {
                return root.right;
            } else if (root.right == null) {
                return root.left;
            } else {
                // Node with two children
                Node succ = getSuccessor(root);
                root.key = succ.key;
                root.right = delNode(root.right, succ.key);
            }
        }
        return root;
    }

    // Search in BST (recursive)
    public static boolean search(Node root, int x) {
        if (root == null)
            return false;
        if (x == root.key)
            return true;
        else if (x < root.key)
            return search(root.left, x);
        else
            return search(root.right, x);
    }

    // Inorder traversal
    public static void inorder(Node root) {
        if (root != null) {
            inorder(root.left);
            System.out.print(root.key + " ");
            inorder(root.right);
        }
    }

    public static void main(String[] args) {
        Node root = new Node(15);
        root.left = new Node(5);
        root.right = new Node(20);
        root.left.left = new Node(3);
        root.right.right = new Node(80);
        root.right.left = new Node(18);
        root.right.left.left = new Node(16);

        int x = 15;
        root = delNode(root, x);

        System.out.println("Inorder traversal after deletion:");
        inorder(root);
    }
}